code/modules/reagents/reagent_containers.dm

```diff
@@ -36,13 +39,139 @@
 /obj/item/weapon/reagent_containers/attackby(obj/item/I as obj, mob/user as mob)
 	return
 */
-/obj/item/weapon/reagent_containers/afterattack(obj/target, mob/user , flag)
+
+/**
+ * This usually handles reagent transfer between containers and splashing the contents.
+ * Please see `transfer()` for a general reusable proc for that.
+ *
+ * If you're wondering why you're splashing machinery that accepts beakers when
+ * inserting them, it's because the machine is returning `FALSE` on `attackby()`,
+ * which causes `afterattack()` to be called. Return 1 instead on those cases.
+ *
+ * If your container is splashing/transferring things at a distance, your `afterattack()`
+ * isn't checking for adjacency. For that, check that `adjacency_flag` is `TRUE`.
+ */
+/obj/item/weapon/reagent_containers/afterattack(var/obj/target, var/mob/user, var/adjacency_flag, var/click_params)
 	return
 
+/**
+ * Transfer reagents between reagent_containers/reagent_dispensers.
+ *
+ * @warning You should typecheck everything before calling this proc!
+ * @warning This doesn't check for open containers!
+ * @warning Do not call this unless you know what you're doing!
+ */
+/proc/transfer_sub(var/atom/source, var/atom/target, var/amount, var/mob/user)
+	if (source:is_empty())

```

Comment:
Don't do this.
If the coder has to typecheck for this proc, you can cast for it.

Better yet, you can cast AND typecheck.